In [25]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, accuracy_score, roc_curve
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
import matplotlib.pyplot as plt

In [10]:
df = pd.read_csv('/Users/ahmadallaou/Documents/UCB Repos/datasci207_project/data/raw/skyserver_sdss/dr16_galaxySpecPhoto.csv')

/var/folders/w9/dx4nvr3j01s9jp6j1fm286vc0000gn/T/ipykernel_36793/649618218.py:1: DtypeWarning: Columns (21,22) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/Users/ahmadallaou/Documents/UCB Repos/datasci207_project/data/raw/skyserver_sdss/dr16_galaxySpecPhoto.csv')


In [11]:
df.columns

Index(['specobjid', 'dr8objid', 'dr7objid', 'ra', 'dec', 'nvote_tot',
       'nvote_std', 'nvote_mr1', 'nvote_mr2', 'nvote_mon', 'p_el', 'p_cw',
       'p_acw', 'p_edge', 'p_dk', 'p_mg', 'p_cs', 'instrument', 'redshift',
       'sdss_class_string', 'sdss_subclass_string', 'class_noqso',
       'subClass_noqso', 'spectroFlux_u', 'spectroFlux_g', 'spectroFlux_r',
       'spectroFlux_i', 'spectroFlux_z', 'spectroSynFlux_u',
       'spectroSynFlux_g', 'spectroSynFlux_r', 'spectroSynFlux_i',
       'spectroSynFlux_z', 'elodieObject', 'elodieSpType', 'elodieBV',
       'elodieTEff', 'elodieLogG', 'elodieFeH', 'elodieZ', 'sdss_class_number',
       'clean', 'u', 'g', 'r', 'i', 'z', 'nObserve', 'nDetect', 'nEdge',
       'score'],
      dtype='object')

In [12]:
features = ['u','g','r','i','z','redshift','ra','dec']

In [13]:
label = ['sdss_class_string']

In [14]:
df['sdss_class_string'].value_counts()

GALAXY    786076
STAR        7328
QSO         7223
Name: sdss_class_string, dtype: int64

In [15]:
X = df[features]
y = df[label]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, stratify=y, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

clf = RandomForestClassifier(n_estimators=100, random_state=42)

clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

/var/folders/w9/dx4nvr3j01s9jp6j1fm286vc0000gn/T/ipykernel_36793/2050606733.py:17: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  clf.fit(X_train, y_train)


0.9931886972342614
              precision    recall  f1-score   support

      GALAXY       0.99      1.00      1.00    235824
         QSO       0.90      0.33      0.49      2167
        STAR       0.98      0.97      0.97      2198

    accuracy                           0.99    240189
   macro avg       0.96      0.77      0.82    240189
weighted avg       0.99      0.99      0.99    240189



In [ ]:
print(accuracy)
print(class_report)

In [23]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

y_pred_knn = knn.predict(X_test)

accuracy_knn = accuracy_score(y_test, y_pred_knn)
class_report_knn = classification_report(y_test, y_pred_knn)

/Users/ahmadallaou/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/Users/ahmadallaou/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


In [24]:
print(accuracy_knn)
print(class_report_knn)

0.9877971097760514
              precision    recall  f1-score   support

      GALAXY       0.99      1.00      0.99    235824
         QSO       0.96      0.09      0.17      2167
        STAR       0.77      0.82      0.79      2198

    accuracy                           0.99    240189
   macro avg       0.90      0.64      0.65    240189
weighted avg       0.99      0.99      0.98    240189



In [26]:
logreg = LogisticRegression(max_iter=10000, random_state=42)
logreg.fit(X_train, y_train)

y_pred_logreg = logreg.predict(X_test)

accuracy_logreg = accuracy_score(y_test, y_pred_logreg)
class_report_logreg = classification_report(y_test, y_pred_logreg)

/Users/ahmadallaou/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [27]:
print(accuracy_logreg)
print(class_report_logreg)

0.9886297873757749
              precision    recall  f1-score   support

      GALAXY       0.99      1.00      0.99    235824
         QSO       0.95      0.08      0.15      2167
        STAR       0.78      0.93      0.85      2198

    accuracy                           0.99    240189
   macro avg       0.91      0.67      0.66    240189
weighted avg       0.99      0.99      0.99    240189



# Classifying the Objects
Using the SDSS dataset, we've trained a model that classifies an object using the features:
-   U:
-   G:
-   R:
-   I:
-   Z:
-   Redshift:
-   ra:
-   dec:

to classify an object as either a galaxy, quasar or star. We have opted to use several models to classify these objects including (with their respective scores):
-   RFC
    -   Accuracy: 99.3%
    -   Recall: 
        -   Galaxy: 100%
        -   QSO: 33%
        -   STAR: 97%
    -   Precision:
        -   Galaxy: 99%
        -   QSO: 90%
        -   STAR: 98%
    -   F1 Score:
        -   Galaxy: 100%
        -   QSO: 49%
        -   STAR: 97%
<br></br>
-   KNN
    -   Accuracy: 98.8%
    -   Recall: 
        -   Galaxy: 100%
        -   QSO: 9%
        -   STAR: 82%
    -   Precision:
        -   Galaxy: 99%
        -   QSO: 96%
        -   STAR: 77%
    -   F1 Score:
        -   Galaxy: 99%
        -   QSO: 17%
        -   STAR: 79%
<br></br>
-   Logistic Regression
    -   Accuracy: 98.9%
    -   Recall: 
        -   Galaxy: 100%
        -   QSO: 8%
        -   STAR: 93%
    -   Precision:
        -   Galaxy: 99%
        -   QSO: 95%
        -   STAR: 78%
    -   F1 Score:
        -   Galaxy: 99%
        -   QSO: 15%
        -   STAR: 85%

